### PR TITLE
Cleanup defines for specialized handle types

### PR DESCRIPTION
**Pull request by jkotas**
- Introduce FEATURE_VARIABLE_HANDLES, FEATURE_REFCOUNTED_HANDLES and FEATURE_WEAK_NATIVE_COM_HANDLES defines to control compilation of support for these specialized handle types.
- Stop compiling FEATURE_VARIABLE_HANDLES and FEATURE_WEAK_NATIVE_COM_HANDLES support into the live runtime since these handle types are not currently used. (The support is still compiled into standalone GC for backward compatibility.)
- Delete some dead code

### CODE REVIEW

File: src/coreclr/gc/objecthandle.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ void CALLBACK VariableTraceDispatcher(_UNCHECKED_OBJECTREF *pObjRef, uintptr_t *
     }
 }
 
-#if defined(FEATURE_COMINTEROP) || defined(FEATURE_COMWRAPPERS) || defined(FEATURE_OBJCMARSHAL) || defined(FEATURE_NATIVEAOT)
```
Comment (jkotas) [inline src/coreclr/gc/objecthandle.cpp:69]: This deletes number of `FEATURE_NATIVEAOT` ifdefs in the GC. `FEATURE_NATIVEAOT` ifdefs in the GC are anti-patern.

File: src/coreclr/gc/gcinterface.h [hunk 2]
```diff
@@ -474,9 +472,8 @@ typedef enum
      * useful when passing a pointer to object innards out of the runtime while GC
      * may be enabled.
```
Comment (teo-tsirpanis) [inline src/coreclr/gc/gcinterface.h:473]: The documentation is identical to `HNDTYPE_PINNED`. Would it be worth it to add some text to differentiate them?

Comment (jkotas) [reply src/coreclr/gc/gcinterface.h]: I have updated the comment. (I expect that we are going to delete the support for async pinned handles at some point - it is not worth it to put much effort into it.)

File: src/coreclr/gc/objecthandle.cpp [hunk 3]
```diff
@@ -424,9 +427,9 @@ void CALLBACK ScanPointerForProfilerAndETW(_UNCHECKED_OBJECTREF *pObjRef, uintpt
         break;
     case    HNDTYPE_WEAK_SHORT:
     case    HNDTYPE_WEAK_LONG:
-#ifdef FEATURE_COMINTEROP
+#ifdef FEATURE_WEAK_NATIVE_COM_HANDLES
     case    HNDTYPE_WEAK_NATIVE_COM:
-#endif // FEATURE_COMINTEROP
+#endif
```
Comment (mangod9) [inline src/coreclr/gc/objecthandle.cpp:432]:
```suggestion
#endif // FEATURE_WEAK_NATIVE_COM_HANDLES
```

File: src/coreclr/gc/objecthandle.cpp [hunk 4]
```diff
@@ -1015,6 +993,7 @@ void TraceVariableHandlesBySingleThread(HANDLESCANPROC pfnTrace, uintptr_t lp1,
         walk = walk->pNext;
     }
 }
+#endif
```
Comment (mangod9) [inline src/coreclr/gc/objecthandle.cpp:996]:
```suggestion
#endif // FEATURE_VARIABLE_HANDLES
```

File: src/coreclr/gc/objecthandle.cpp [hunk 5]
```diff
@@ -439,29 +442,27 @@ void CALLBACK ScanPointerForProfilerAndETW(_UNCHECKED_OBJECTREF *pObjRef, uintpt
...
+            }
+            if ((nVarHandleType & VHT_PINNED) != 0)
+            {
+                rootFlags |= kEtwGCRootFlagsPinning;
+            }
 
-        // No special ETW flag for strong handles (VHT_STRONG)
-    }
-#else
-        _ASSERTE(!"Variable handle encountered");
```
Comment (mangod9) [inline src/coreclr/gc/objecthandle.cpp:460]: assume this assert is no longer necessary?

Comment (jkotas) [reply src/coreclr/gc/objecthandle.cpp:460]: It can be moved to the `default:` case.

**Comment by cshung** [comment]
I have found some more opportunities for cleanup here https://github.com/cshung/runtime/commit/590eae65a5e79df0ab17fa68a4f4e7327c6e1731, can you please take a look?

**Comment by jkotas** [comment]
> I have found some more opportunities for cleanup here https://github.com/cshung/runtime/commit/590eae65a5e79df0ab17fa68a4f4e7327c6e1731, can you please take a look?

Thank you. I have pushed a commit inspired by your change:
- Deleted runtime support for handles that the runtime no longer uses instead of ifdefing it.
- I have omitted the change that added the `FEATURE_...HANDLE` ifdef into the contract itself. The `FEATURE_...HANDLE` ifdefs have applied to the implementation only and ifdefs do not work well in contracts.

**Review by mangod9 — APPROVED** [review]

**Pull request merged**